themes/default/content/blog/next-level-iac-pulumi-runtime-logic/index.md

```diff
@@ -0,0 +1,428 @@
+
+## Fetch data to generate resources dynamically
+
+To build on the static-website scenario, suppose you wanted to retrieve some data from an external source and then use it to provision a set of cloud resources dynamically. For instance, you might want to let your marketing team create their own short URLs for social media posts. For this, you might once again reach for Bash to pull those URLs from a corporate CMS, either before or after running `pulumi` itself, to create the corresponding redirects imperatively using `curl` and the `aws` CLI.
+
+Again, totally reasonable. But why bother, when you have the full capability of {{< langhost >}} at fingertips? Just add a few lines to the program to fetch the data and be on your way:
+
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript
+// ...
+
+// Fetch some redirects from a hypothetical CMS, pulling its URL from the environment.
+const redirects = fetch(`${process.env.CMS_ENDPOINT}/redirects.json`)
+    .then(response => response.json())
+    .then(items => items.forEach((redirect: any, i: number) => {
+
+        // Create an S3 website redirect for each one.
+        new aws.s3.BucketObject(`redirect-${i}`, {
+            bucket: bucket.id,
+            key: redirect.from,
+            websiteRedirect: redirect.to,
+            acl: "public-read",
+        });
+    }),
+);
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python
+#...
+import json
+import os
+import requests
+
+# ...
+
+# Fetch some redirects from a hypothetical CMS.
+response = requests.get(f"{os.environ['CMS_ENDPOINT']}/redirects.json")
+redirects = json.loads(response.text)
+
+# Create an S3 website redirect for each one.
+for i, redirect in enumerate(redirects):
+    aws.s3.BucketObject(f"redirect-{i}",
+        bucket=bucket.id,
+        key=redirect["from"],
+        website_redirect=redirect["to"],
+        acl="public-read",
+    )
+```
+
+{{% /choosable %}}
+
+Notice the example parses the response using the language's built-in JSON support. If your CMS happened to return data in some other format, or required some additional credentials for authorization, no problem --- your language and its ecosystem are already there to help you.
+
+Also notice the example reads the source URL directly from the environment. This is typically more cumbersome (or even impossible) with other tools, requiring that you name your environment variables in specific ways, and then tack on additional blocks of configuration in order to read and assign them. But as you can see, since you're working with {{< langhost >}}, it's as easy as referencing a regular variable.
+
+Finally, notice it also uses a familiar looping construct. That's next!
+
+## Write clear, simple loops and conditionals
```

Comment:
this one seems different than the others. the others are pre and post processing using languages. this one i think should be broken out into a completely separate blog post where we can dive into the glories of programming language constructs